lib/mac_layer/mac.cpp

```diff
@@ -97,98 +101,180 @@ namespace ex2 {
+        // header (we set to 0 always), we can check it and catch cases where
+        // the Golay decoder provides a false positive (i.e., when there are
+        // more than 4 bit errors in 12 bits.
+        //
+        // When there is a false positive, we can't trust anything about the MPDU.
+        //
+        // If we have the first MPDU fragment already, this might be the MPDU
+        // we are expecting. If it's not the last fragment, we can wait for
+        // another to arrive and let the logic below pad things out. If it is
+        // the last fragment, either a new packet will send a new first MPDU
+        // fragement and the current one will complete, or it will timeout
+        // and complete. Either way, here we do nothing, just return.
+        //
+        // If we don't have the first MPDU fragement already, this might be it,
+        // but we can't trust it. So we ignore it.
+        if (mpdu.getMpduHeader()->getUserPacketFragmentIndex() != MPDU_HEADER_USER_PACKET_FRAGMENT_INDEX_DEFAULT) {
+#if MAC_DEBUG
+            printf("Received an MPDU with user packet fragment index not zero\n");
+#endif
+          return MAC_UHFPacketProcessingStatus::READY_FOR_NEXT_UHF_PACKET;
+        }
 
-        // @todo Could do some header checks in case we are unlucky. Check packet length is >= 0, for example
+        // A common reason making an MPDU will fail and throw an exception is
+        // that the packet header information is corrupt. We rely on the
+        // MPDUHeader class to check that, including that the error correction
+        // scheme in the header is consistent with the current one in use.
 
-        // The first MPDU transmitted may have a header in it, which is
-        // necessary to make the full packet. Without it, there is no point
-        // in making an application packet so we keep track of its reception.
+        // If we already have the first MPDU, we are looking for more...
         if (m_firstFragmentReceived) {
           // If we are here, we expected more raw MPDUs.
 
+          // We already have the user packet length and since it's in every
+          // MPDU header, let's check to see if the current one is consistent
+          // with the first one we received.
+          if (mpdu.getMpduHeader()->getUserPacketPayloadLength() != m_currentPacketLength) {
+            // Ooops, we must have missed all the MPDUs that match the first
+            // MPDU fragment. This is similar to the scenario where we have the
+            // first fragment, then get another first fragment. Best we can do is
+            // dump the current packet processing.
+#if MAC_DEBUG
+            printf("Received an MPDU with user packet length that does not match the current length\n");
+#endif
+            m_firstFragmentReceived = false;
+            m_mpduCodewordFragmentCount = 0;
+            // if it happens to be the first fragment, then let's start processing for a new packet.
```

Comment:
I think this refers to the line that follows where we check if the current MPDU has codeword fragment index == 0...

No, it can definitely be incorrect, but in any MPDUheader the fragment count could be bad, which is why we do other checks (out of order, more than expected given the user packet size, etc). 

Since we agree that we will try our best to produce a packet based on received MPDUs, even if they might be suspect, we need to rely on higher-than-MAC level checking, too. Make sense?

Reply:
Put another way, if the codeword fragment index is correct, then we have to start receiving the new packet, right? Not doing that means we definitely miss an MPDU, so I think this is the right thing to do.

Reply:
Sounds good